doc/source/rllib/rllib-fault-tolerance.rst

```diff
@@ -0,0 +1,74 @@
+.. include:: /_includes/rllib/announcement.rst
+
+.. include:: /_includes/rllib/we_are_hiring.rst
+
+Fault Tolerance And Elastic Training
+======================================================
+
+RLlib handles common failures modes (e.g. machine failures, spot isntance preemption,
+network outages, or Ray cluster failures) via logics at multiple system levels.
+This doc discusses them in details.
+
+Fault Tolerance and Recovery Provided by Ray Tune
+-------------------------------------------------
+
+Ray Tune provides fault tolerance and recovery at trial level.
+
+Majority of RLlib runs are launched with Ray Tune, these trials periodically checkpoint
+the entire training states to user specified persistent storage locations. If a trial fails,
+Ray Tune will automatically restart it from the latest
+:ref:`checkpointed <tune-two-types-of-ckpt>` state.
+
+This serves as our last line of defense, catching any unexpected errors that are not
+handled by RLlib.
+
+Elastic RolloutWorkers
+----------------------
+
+RLlib supports self-recovering and elastic workersets for both rollout and evaluation workers.
+This provides fault tolerance at worker level.
+
+Worker fault tolerance can be turned on by setting config ``recreate_failed_workers`` to True.
+Under the hood, RLlib relies on Ray Core :ref:`actor fault tolerance <actor-fault-tolerance>`
+to automatically recover failed worker actors. Whenever a worker actor is reconstructed,
+RLlib will make sure its latest state gets synced before new episodes can be sampled. 
+
+RLlib workers are also elastic meaning that they do not require the full set of workers to
+function. For example, if an RLlib trial uses spot instances, it is quite possible for the
+trial to only have partial resources at a given time, potentially resulting in a subset
+of workers not being able to get scheduled. In this case, RLlib will make sure training and
+evaluation carries on with whatever healthy workers left at a reduced speed.
+RLlib achieves this by utilizing a
+`state-aware and fault tolerant actor manager <https://github.com/ray-project/ray/blob/master/rllib/utils/actor_manager.py>`__.
+
+Env Fault Tolerance
+-------------------
+
+In addition to worker fault tolerance, RLlib offers fault tolerance at environment level as well.
+
+It's common for a rollout or evaluation workers to run multiple environments in parallel to take
+advantage of, for example, the parallel computing power that GPU offers. It may then be wasteful
```

Comment:
the gpu thing is not so much related to parallel workers as leverging many cpu workers is.  So for example instead of saying `take advantage of, for example, the parallel computing power that GPU offers` say  `take advantage of, cpu cores in a compute cluster`

Reply:
wait wait, this is actually specifically for taking advantage of the GPUs, which has much higher parallel computing power.
like if all we have are CPUs, there is little reason to run multiple envs in a single workers. we could just run multiple workers.
if you have multiple envs, obs from them are inferenced together as a single batch in env_runner.

Reply:
yeah I think I was confused about what we are talking about here. This means that the documentation should be clarified on num_envs_per_worker or num_workers?